cli/src/api/endpoints.rs

```diff
@@ -1,84 +1,299 @@
+    url.path_segments_mut()
+        .unwrap()
+        .pop_if_empty()
+        .extend(["job", "projects", "name", pkg_id]);
+    Ok(url)
 }
 
 /// GET /job/projects/overview
-pub fn get_project_summary(api_uri: &str) -> String {
-    format!("{api_uri}/{API_PATH}/job/projects/overview")
+pub fn get_project_summary(api_uri: &str) -> Result<Url, BaseUriError> {
+    Ok(get_api_path(api_uri)?.join("job/projects/overview")?)
 }
 
 /// PUT /job/projects
-pub fn put_create_project(api_uri: &str) -> String {
-    format!("{api_uri}/{API_PATH}/job/projects")
+pub fn put_create_project(api_uri: &str) -> Result<Url, BaseUriError> {
+    Ok(get_api_path(api_uri)?.join("job/projects")?)
 }
 
 /// GET /settings/current-user
-pub(crate) fn get_user_settings(api_uri: &str) -> String {
-    format!("{api_uri}/{API_PATH}/settings/current-user")
+pub(crate) fn get_user_settings(api_uri: &str) -> Result<Url, BaseUriError> {
+    Ok(get_api_path(api_uri)?.join("settings/current-user")?)
 }
 
 /// PUT /settings/current-user
-pub(crate) fn put_user_settings(api_uri: &str) -> String {
-    format!("{api_uri}/{API_PATH}/settings/current-user")
+pub(crate) fn put_user_settings(api_uri: &str) -> Result<Url, BaseUriError> {
+    Ok(get_api_path(api_uri)?.join("settings/current-user")?)
 }
 
-/// GET /groups/list
-pub(crate) fn group_list(api_uri: &str) -> String {
-    format!("{api_uri}/{API_PATH}/groups")
+/// GET /groups
+pub(crate) fn group_list(api_uri: &str) -> Result<Url, BaseUriError> {
+    Ok(get_api_path(api_uri)?.join("groups")?)
 }
 
-/// POST /groups/create
-pub(crate) fn group_create(api_uri: &str) -> String {
-    format!("{api_uri}/{API_PATH}/groups")
+/// POST /groups
+pub(crate) fn group_create(api_uri: &str) -> Result<Url, BaseUriError> {
+    Ok(get_api_path(api_uri)?.join("groups")?)
 }
 
 /// GET /groups/<groupName>/projects
-pub fn group_project_summary(api_uri: &str, group: &str) -> String {
-    format!("{api_uri}/{API_PATH}/groups/{group}/projects")
+pub fn group_project_summary(api_uri: &str, group: &str) -> Result<Url, BaseUriError> {
+    let mut url = get_api_path(api_uri)?;
+    url.path_segments_mut()
+        .unwrap()
+        .pop_if_empty()
+        .extend(["groups", group, "projects"]);
+    Ok(url)
 }
 
 /// GET /.well-known/openid-configuration
-pub fn oidc_discovery(api_uri: &str) -> String {
-    format!("{api_uri}/{API_PATH}/.well-known/openid-configuration")
+pub fn oidc_discovery(api_uri: &str) -> Result<Url, BaseUriError> {
+    Ok(get_api_path(api_uri)?.join(".well-known/openid-configuration")?)
+}
+
+fn parse_base_url(api_uri: &str) -> Result<Url, BaseUriError> {
+    let mut url = Url::parse(api_uri)?;
+
+    // Ensure the path can be a base and ends with a slash so it can be safely joined to.
+    // If we don't do this, https://example.com/a and https://example.com/a/ are different.
+    url.path_segments_mut()
+        .map_err(|_| ParseError::RelativeUrlWithCannotBeABaseBase)?
```

Comment:
`ParseError::RelativeUrlWithCannotBeABaseBase` what does this error variant mean?

Reply:
It's if the URL is something like `data:text…` or `mailto:user@example.com` etc. https://docs.rs/url/2.2.2/url/struct.Url.html#method.cannot_be_a_base